In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import random

In [2]:
# default cleaning method until proven otherwise

def clean_census_frame(csv_path , head=False , reset=True , set_index=False ):
    '''
    inputs) 
        >> csv_path
            > path to csv
        >> head
            > default=False
                >> if != False
                    > integer
                        >> returns the first {head} rows (using .head() method) 
                            > instead of enitre dataframe
        >> reset
            > default=True
                >> resets index after taking out rows
            > if set to False
                >> will not reset index
        >> set_index
            > default=False
            > if != False
                >> will set_index of new df to set_index
    output)
        >> dataframe cleaned like 2000 Census age&sex by 5-digit Zip Code (also how 2010 for same is cleaned)
    how)
        1. reads in csv , assumes it's large
        2. makes a copy for editing 
            > and potential future use
        3. locates readable column names  and non-readable names 
            > readable
                    > e.g. Estimate; SEX AND AGE - Total population
                >> assumes they are currently in row 0
            > non-readable
                    > e.g. HC01_VC03
                >> assumes they are currently == dataframe.columns
        4. replaces dataframe.columns (non-readable) with readable column names
            > and drops the old 0th column (column where readable names were stored)
        
    '''
    # load data
    df = pd.read_csv( csv_path , low_memory=False )

    # and copy
    _df = df.copy()

    # reset column names to current 0th row values
    _df.columns = _df.iloc[0]
    # new 2000 dataframe without row where values are from
    clean_df = _df[1:]
    
    # default
    if reset==True:
        # reset index
        clean_df = clean_df.reset_index()
        
    # set_index
    if set_index:
        clean_df = clean_df.set_index(set_index)
    
    if head:
        # return first {head} rows of dataframe
        return clean_df.head(head)
    else:
        # return dataframe
        return clean_df
    
    
def show_lengths(collection):
    '''
    shows lengths of collection of items
    '''
    # collect
    lengths = []
    # iterate
    for i in collection:
        # add len to collect
        lengths.append(len(i))
    # return all lens
    return lengths

In [ ]:
# test = pd.read_csv('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv',low_memory=False)
# ID=test.copy()
# for _ in range(len(ID.columns)):
#     for i in range(len(ID[ID.columns[_]])):
#         if isinstance(ID[ID.columns[_]][i],str):
#             pass
#         else:
#             print(f'column {_}\nrow {i}\ntype {type(ID[ID.columns[_]][i])}\n{ID[ID.columns[_]][i]}\n\n') 

- ***notes***:
    - every instance is string (was no output)
    - merged & hashed as to avoid running again on accident or restart^

In [ ]:
# reload a year for base comps
t000 = pd.read_csv('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv',low_memory=False)
# limit size
# test = test.head(1000)

In [ ]:
test = clean_census_frame('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv',reset=False)

In [ ]:
t000.columns  #.describe()

In [ ]:
test.columns  #describe()

In [3]:
'''load data'''
# 2011 
y2k11 = clean_census_frame('../data/acs/aff_download/ACS_11_5YR_DP05_with_ann.csv',reset=False)
# 2012
y2k12 = clean_census_frame('../data/acs/aff_download/ACS_12_5YR_DP05_with_ann.csv',reset=False)
#2013
y2k13 = clean_census_frame('../data/acs/aff_download/ACS_13_5YR_DP05_with_ann.csv',reset=False)
# 2014
y2k14 = clean_census_frame('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv',reset=False)
# 2015
y2k15 = clean_census_frame('../data/acs/aff_download/ACS_15_5YR_DP05_with_ann.csv',reset=False)
#2016
y2k16 = clean_census_frame('../data/acs/aff_download/ACS_16_5YR_DP05_with_ann.csv',reset=False)
#2017
y2k17 = clean_census_frame('../data/acs/aff_download/ACS_17_5YR_DP05_with_ann.csv',reset=False)

In [4]:
'''copy data for editing and extracting info'''
# 2011
y11 = y2k11.copy()
# 2012
y12 = y2k12.copy()
# 2013
y13 = y2k13.copy()
# 2014
y14 = y2k14.copy()
# 2015
y15 = y2k15.copy()
# 2016
y16 = y2k16.copy()
# 2017
y17 = y2k17.copy()

In [6]:
'''identify columns'''
# 2011
tags11 = y11.columns  
# 2012
tags12 = y12.columns  
#2013
tags13 = y13.columns  
# 2014
tags14 = y14.columns  
# 2015
tags15 = y15.columns  
#2016
tags16 = y16.columns  
# 2017
tags17 = y17.columns 

In [7]:
'''identify common columns'''
# collection of columns appearing in all 7 dataframes 2011-2017
common_tags = [tag for tag in tags17 if tag in tags11 & tags12 & tags13 & tags14 & tags15 & tags16]

In [52]:
'''collect columns not coexisting in all dataframes from each dataframe'''
# 2011
uncommon_11 = [tag for tag in y11.columns if tag not in common_tags]
# 2012
uncommon_12 = [tag for tag in y12.columns if tag not in common_tags]
# 2013
uncommon_13 = [tag for tag in y13.columns if tag not in common_tags]
# 2014
uncommon_14 = [tag for tag in y14.columns if tag not in common_tags]
# 2015
uncommon_15 = [tag for tag in y15.columns if tag not in common_tags]
# 2016
uncommon_16 = [tag for tag in y16.columns if tag not in common_tags]
# 2017
uncommon_17 = [tag for tag in y17.columns if tag not in common_tags]

In [71]:
"""drop each frame's uncommon columns, reset index"""
# 2011
a_t11 = y11.drop(uncommon_11,axis=1).reset_index()
# # 2012
a_t12 = y12.drop(uncommon_12,axis=1).reset_index()
# # 2013
a_t13 = y13.drop(uncommon_13,axis=1).reset_index()
# # 2014
a_t14 = y14.drop(uncommon_14,axis=1).reset_index()
# # 2015
a_t15 = y15.drop(uncommon_15,axis=1).reset_index()
# # 2016
a_t16 = y16.drop(uncommon_16,axis=1).reset_index()
# # 2017
a_t17 = y17.drop(uncommon_17,axis=1).reset_index()

In [72]:
a_t11.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 220 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(219)
memory usage: 55.6+ MB


In [73]:
a_t12.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 220 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(219)
memory usage: 55.6+ MB


In [74]:
a_t13.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 212 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(211)
memory usage: 53.6+ MB


In [75]:
a_t14.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 212 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(211)
memory usage: 53.6+ MB


In [76]:
a_t15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 212 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(211)
memory usage: 53.6+ MB


In [77]:
a_t16.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 212 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(211)
memory usage: 53.6+ MB


In [78]:
a_t17.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33120 entries, 0 to 33119
Columns: 212 entries, index to Percent Margin of Error; HISPANIC OR LATINO AND RACE - Total housing units
dtypes: int64(1), object(211)
memory usage: 53.6+ MB


In [46]:
print(len(y11.Geography),len(common_tags),len(t11.iloc[1]))
print(len(a_t11.Geography),len(common_tags),len(a_t11.iloc[1]))

33120 211 235
33120 211 219


In [29]:
t11

0                  Id    Id2    Geography  \
1      8600000US00601  00601  ZCTA5 00601   
2      8600000US00602  00602  ZCTA5 00602   
3      8600000US00603  00603  ZCTA5 00603   
4      8600000US00606  00606  ZCTA5 00606   
5      8600000US00610  00610  ZCTA5 00610   
6      8600000US00612  00612  ZCTA5 00612   
7      8600000US00616  00616  ZCTA5 00616   
8      8600000US00617  00617  ZCTA5 00617   
9      8600000US00622  00622  ZCTA5 00622   
10     8600000US00623  00623  ZCTA5 00623   
11     8600000US00624  00624  ZCTA5 00624   
12     8600000US00627  00627  ZCTA5 00627   
13     8600000US00631  00631  ZCTA5 00631   
14     8600000US00637  00637  ZCTA5 00637   
15     8600000US00638  00638  ZCTA5 00638   
16     8600000US00641  00641  ZCTA5 00641   
17     8600000US00646  00646  ZCTA5 00646   
18     8600000US00647  00647  ZCTA5 00647   
19     8600000US00650  00650  ZCTA5 00650   
20     8600000US00652  00652  ZCTA5 00652   
21     8600000US00653  00653  ZCTA5 00653   
22     8600000US00656  00656  ZCTA5 00656   
23     8600000US00659  00659  ZCTA5 00659   
24     8600000US00660  00660  ZCTA5 00660   
25     8600000US00662  00662  ZCTA5 00662   
26     8600000US00664  00664  ZCTA5 00664   
27     8600000US00667  00667  ZCTA5 00667   
28     8600000US00669  00669  ZCTA5 00669   
29     8600000US00670  00670  ZCTA5 00670   
30     8600000US00674  00674  ZCTA5 00674   
...               ...    ...          ...   
33091  8600000US99786  99786  ZCTA5 99786   
33092  8600000US99788  99788  ZCTA5 99788   
33093  8600000US99789  99789  ZCTA5 99789   
33094  8600000US99790  99790  ZCTA5 99790   
33095  8600000US99791  99791  ZCTA5 99791   
33096  8600000US99801  99801  ZCTA5 99801   
33097  8600000US99820  99820  ZCTA5 99820   
33098  8600000US99824  99824  ZCTA5 99824   
33099  8600000US99825  99825  ZCTA5 99825   
33100  8600000US99826  99826  ZCTA5 99826   
33101  8600000US99827  99827  ZCTA5 99827   
33102  8600000US99829  99829  ZCTA5 99829   
33103  8600000US99830  99830  ZCTA5 99830   
33104  8600000US99832  99832  ZCTA5 99832   
33105  8600000US99833  99833  ZCTA5 99833   
33106  8600000US99835  99835  ZCTA5 99835   
33107  8600000US99836  99836  ZCTA5 99836   
33108  8600000US99840  99840  ZCTA5 99840   
33109  8600000US99841  99841  ZCTA5 99841   
33110  8600000US99901  99901  ZCTA5 99901   
33111  8600000US99903  99903  ZCTA5 99903   
33112  8600000US99918  99918  ZCTA5 99918   
33113  8600000US99919  99919  ZCTA5 99919   
33114  8600000US99921  99921  ZCTA5 99921   
33115  8600000US99922  99922  ZCTA5 99922   
33116  8600000US99923  99923  ZCTA5 99923   
33117  8600000US99925  99925  ZCTA5 99925   
33118  8600000US99926  99926  ZCTA5 99926   
33119  8600000US99927  99927  ZCTA5 99927   
33120  8600000US99929  99929  ZCTA5 99929   

0     Estimate; SEX AND AGE - Total population  \
1                                        18533   
2                                        41930   
3                                        54475   
4                                         6386   
5                                        29111   
6                                        70541   
7                                        10617   
8                                        24458   
9                                         5419   
10                                       45314   
11                                       26439   
12                                       35260   
13                                        1836   
14                                       25511   
15                                       19166   
16                                       30987   
17                                       37942   
18                                        5605   
19                                       15359   
20                                        3378   
21                                       14048   
22                                       21684   
23                                       41843   
24       

In [ ]:

test_001 = [tag for tag in tags13 if tag in tags11 & tags12]
test_002 = [tag for tag in tags14 if tag in test_001]
# test_003 = [tag for tag in tags13 if tag in test_001 & tags14]

'''with common columns only'''
# 2011
t11 = t11.drop(common_tags,axis=1)

In [ ]:
len(test_000),len(test_001),len(test_002)#,len(test_003)

In [ ]:
'''# common columns
common_tags = [tag for tag in tags13 if tag in tags11 & tags12]
# chk chg
len(common_tags),len(tags13)
(219, 327)'''

In [ ]:
plz = t11.copy().drop(common_tags,axis=1)
plz

In [ ]:
test11_to_common = t11.copy()

# for u in test11_to_common.columns:
#     if u not in common_tags:
#         test11_to_common = test11_to_common[[u]]   # test11_to_common[u]
# test11_to_common.drop([common_tags],axis=1)
# type(common_tags[1])
# test11_to_common['Id']
type(tuple(common_tags))

In [ ]:
len(tuple(common_tags))

In [ ]:
common_tags[:3]

In [ ]:
# t = t11.copy().drop(['Id', 'Id2', 'Geography'],axis=1)
# t


# for _ in common_tags[1]:
#     tags11 = tags11[_]

In [ ]:
t11_two19  # common_tags

In [ ]:
# len(ID[col_names[_]]) 
# ID[col_names[_]][33120] == ID[col_names[_]][33120]

In [ ]:
# # reload a year for base comps
# test = pd.read_csv('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv',low_memory=False)
# # limit size
# # test = test.head(1000)

In [ ]:
# confirm location of column names and current column names
test.iloc[0]

In [ ]:
# examine metadata (2014 is same as 2011-2013 as same len(.columns))
meta_fourteen_ =  pd.read_csv('../data/acs/aff_download/ACS_14_5YR_DP05_metadata.csv')
for _ in range(1,len(meta_fourteen_)):
    print(str(meta_fourteen_.Id[_-1:_])+'\n')

- ***notes***:
    - looks nothing like the online American Fact Finder version 
        - always fun
    - online 2014
        - highest columns are Geography
            - ZCTA5 00601
            - which is broken into subcolumns
                - Total, Male, Female
                - each of which are broken into subcolumns
                    - Estimate, Margin of Error
        - effectively
            - each Geography is represented by 6 columns
        - seems I have more information regarding race and ethnicity per zip
- ***actions***:
    - represent each area by Geography in each table
    - compare evolution of that Geography
        - Clustering (KNN)
        - Regression

In [ ]:
"""
# 2011
y2k11 = clean_census_frame('../data/acs/aff_download/ACS_11_5YR_DP05_with_ann.csv',reset=False)
# 2012
y2k12 = clean_census_frame('../data/acs/aff_download/ACS_12_5YR_DP05_with_ann.csv',reset=False)
#2013
y2k13 = clean_census_frame('../data/acs/aff_download/ACS_13_5YR_DP05_with_ann.csv',reset=False)
# 2014
y2k14 = clean_census_frame('../data/acs/aff_download/ACS_14_5YR_DP05_with_ann.csv')
# 2015
y2k15 = clean_census_frame('../data/acs/aff_download/ACS_15_5YR_DP05_with_ann.csv')
# 2016
y2k16 = clean_census_frame('../data/acs/aff_download/ACS_16_5YR_DP05_with_ann.csv')
# 2017
y2k17 = clean_census_frame('../data/acs/aff_download/ACS_17_5YR_DP05_with_ann.csv')
"""

In [ ]:
# as it's always come with such joy, let's check length
full_yrs = [y2k11,y2k12,y2k13,y2k14,y2k15,y2k16,y2k17]
# we want these to all be the same
show_lengths(full_yrs)

In [ ]:
# is Geography consistent?

# a = sum(y2k11.Geography != y2k13.Geography)  
# b = sum(y2k12.Geography != y2k14.Geography) 
# c = sum(y2k13.Geography != y2k15.Geography)
# d = sum(y2k14.Geography != y2k16.Geography)
# e = sum(y2k15.Geography != y2k17.Geography)
# f = sum(y2k16.Geography != y2k11.Geography)

# print(sum([a,b,c,d,e,f]))


In [ ]:
# find common columns across all frames
_11_cols = y2k11.columns
_12_cols = y2k12.columns
_13_cols = y2k13.columns
_14_cols = y2k14.columns
_15_cols = y2k15.columns
_16_cols = y2k16.columns
_17_cols = y2k17.columns

collection = [_11_cols,_12_cols,_13_cols,_14_cols,_15_cols,_16_cols,_17_cols]

for _ in range(len(collection)):
    a = len([col for col in collection[_] if col in y2k11])
    b = len([col for col in collection[_] if col in y2k12])
    c = len([col for col in collection[_] if col in y2k13])
    d = len([col for col in collection[_] if col in y2k14])
    e = len([col for col in collection[_] if col in y2k15])
    f = len([col for col in collection[_] if col in y2k16])
    g = len([col for col in collection[_] if col in y2k17])
    out = [_,a,b,c,d,e,f,g]
    for i in out:
        print(i)
    print('\n')

- ***notes***:
    - use 2011 or 2012 columns to find common
    - max len shared columns == 220
- ***action***:
    - find shared columns

In [ ]:
# a = sum(y2k11.Geography != y2k13.Geography)  
# b = sum(y2k12.Geography != y2k14.Geography) 
# c = sum(y2k13.Geography != y2k15.Geography)
# d = sum(y2k14.Geography != y2k16.Geography)
# e = sum(y2k15.Geography != y2k17.Geography)
# f = sum(y2k16.Geography != y2k11.Geography)

# geo11 = y2k11.Geography
# geo12 = y2k12.Geography
# geo13 = y2k13.Geography
# geo14 = y2k14.Geography
# geo15 = y2k15.Geography
# geo16 = y2k16.Geography
# geo17 = y2k17.Geography

In [ ]:
# # zip codes shared across all dataframes, index
# frames = []
# # 2011 zips
# # watch out for 0,1 flipflop due to adjusting sorts above
# for _ in range(len(geo11)): 
#     # set geo11
#     g = geo11[_]
#     # 2012 zips
#     if g == geo12[_]:
#         # 2013 zips
#         if g == geo13[_]:
#             # 2014 zips
#             if g == geo14[_]:
#                 # 2015 zips
#                 if g == geo15[_]:
#                     # 2016 zips
#                     if g == geo16[_]:
#                         # 2017 zips
#                         if g == geo17[_]:
#                             frames.append(_)
# # make sure frames are good to set_index
# if len(frames) != len(geo11) -1:
#     raise Exception(f'len(frames) != len(geo11)\n{len(frames)} != {len(geo11)}')
# determine shared columns across all dataframes, used to ensure codes across time/geo are comparable
columns = []
# 2011 columns
for _ in y2k11.columns:
    # 2012 columns
    if _ in y2k12.columns:
        # 2013 columns
        if _ in y2k13.columns:
            # 2014 columns
            if _ in y2k14.columns:
                # 2015 columns
                if _ in y2k15.columns:
                    # 2016 columns
                    if _ in y2k16.columns:
                        # 2017 columns
                        if _ in y2k17.columns:
                            columns.append(_)
# adjust dataframes to coexistance only
# for _ in columns:


In [ ]:
# _11_cols = y2k11.copy().columns
# _12_cols = y2k12.copy().columns
# _13_cols = y2k13.copy().columns
# _14_cols = y2k14.copy().columns
# _15_cols = y2k15.copy().columns
# _16_cols = y2k16.copy().columns
_17_cols = y2k17.copy().columns

drop_17 = [col for col in _17_cols if col not in columns]

In [ ]:
len(_17_cols)-len(drop_17), len(columns)

In [ ]:
# y2k11[[columns]]

In [ ]:
y11 = y2k11.copy()
y12 = y2k12.copy()
y13 = y2k13.copy()
y14 = y2k14.copy()
y15 = y2k15.copy()
y16 = y2k16.copy()
y17 = y2k17.copy()

In [ ]:
frames = [y11,y12,y13,y14,y15,y16,y17]

In [ ]:
coc = pd.concat(frames,sort=True,axis=1)

pd.to_numeric(coc)

In [ ]:
"""we're going to make a bunch of small dataframes for each zip code
        this should proove useful as zip code is common across all these frames and many other Census datasets"""
# pandas.core.series.Series
# titles = y2k11.Geography

# set years index
idx = ['2011','2012','2013','2014','2015','2016','2017']

blank_frames = []

In [ ]:
y2k14

In [ ]:
'''MAKE IT FASTER >> done to speed up modeling process'''

# 2011
sy2k11 = y2k11.head(1000)
# 2012
sy2k12 = y2k12.head(1000)
#2013
sy2k13 = y2k13.head(1000)
# 2014
sy2k14 = y2k14.head(1000)
# 2015
sy2k15 = y2k15.head(1000)
# 2016
sy2k16 = y2k16.head(1000)
# 2017
sy2k17 = y2k17.head(1000)

In [ ]:
sy2k14

In [ ]:
# gather column names
column_names = [_ for _ in sy2k11[:0]]
column_names

In [ ]:
for _ in [sy2k11,sy2k12,sy2k13,sy2k14,sy2k15,sy2k16,sy2k17]:
    print(_.info(),'\n\n')

In [ ]:
sy2k14

In [ ]:
# determine columns in 2015, 2016 and 2017 which are not seen in 2011-2014 
eleven_columns = [name for name in sy2k11.columns]
twelve_columns = [name for name in sy2k12.columns]
thirteen_columns = [name for name in sy2k12.columns]
fourteen_columns = [name for name in sy2k12.columns]
fifteen_columns = [name for name in sy2k15.columns]
sixteen_columns = [name for name in sy2k16.columns]
seventeen_columns = [name for name in sy2k17.columns]

In [ ]:
# compare lengths of columns
show_lengths([eleven_columns,twelve_columns,thirteen_columns,fourteen_columns,fifteen_columns,sixteen_columns,seventeen_columns])

In [ ]:
'''make sure those with 328 columns all have the same columns'''

# collect similar
w_328_cols = [eleven_columns,twelve_columns,thirteen_columns,fourteen_columns]

# count non coexist 
same_len_non_coexist = 0

# show lens
show_w_328 = show_lengths(w_328_cols)
# double check lengths
for i in range(len(show_w_328)):
    if show_w_328[i] != show_w_328[-i]:
        raise Exception(f'ERROR len != len\n{show_w_328[i]} != {show_w_328[-i]}')

# compare all to initial 
for name in eleven_columns: 
    if name not in twelve_columns:
        same_len_non_coexist+=1
        # raise Exception(f'FLAWED ASSUMPTION if name in y2k12.columns name = {name}')
    elif name not in thirteen_columns:
        same_len_non_coexist+=1
        #raise Exception(f'FLAWED ASSUMPTION if name in y2k13.columns name = {name}')
    elif name not in fourteen_columns:
        same_len_non_coexist+=1
        #raise Exception(f'FLAWED ASSUMPTION if name in y2k14.columns name = {name}')

# final hurdle, check for non coexisting (based on prior hoops this should be a given)
if same_len_non_coexist > 0:
    raise Exception(f'FLAWED ASSUMPTION same_len_non_coexist ({same_len_non_coexist}) > 0')

In [ ]:
'''make sure those with 340 columns all have the same columns'''

for _ in range(len(fifteen_columns)): 
    if fifteen_columns[_] != sixteen_columns[_]:
        raise Exception(f'FLAWED ASSUMPTION if name in y2k12.columns name = {name}')

In [ ]:
def col_name_eda():
    '''see if those 328 column names are seen in 2015, 2016, and 2017'''
    # store non coexist 2011 vs 2015
    non_coexist_2015 = []
    # store non coexist 2011 vs 2016
    non_coexist_2016 = []
    # store non coexist 2011 vs 2017
    non_coexist_2017 = []

    # compare to initial 
    for _ in range(len(eleven_columns)): 
        # 2017
        if eleven_columns[_] not in seventeen_columns:
            # update 2017 bag w/ index & instance
            non_coexist_2017.append((_,eleven_columns[_]))
        # 2016
        if eleven_columns[_] not in sixteen_columns:
            # update 2016 bag w/ index & instance
            non_coexist_2016.append((_,eleven_columns[_]))
        # 2015
        if eleven_columns[_] not in fifteen_columns:
            # update 2015 bag w/ index & instance
            non_coexist_2015.append((_,eleven_columns[_]))


    '''reverse vs 2016 and then vs 2017 if 2015 == 2016'''
    if fifteen_columns == sixteen_columns:

        # store non coexist 2016 vs 2011
        _from_2016_non_coexist_2011 = []
        # store non coexist 2017 vs 2011
        _from_2017_non_coexist_2011 = []

        # 2017
        for _ in range(len(seventeen_columns)):
            if seventeen_columns[_] not in eleven_columns:
                # update 2017 bag w/ index & instance
                _from_2017_non_coexist_2011.append((_,seventeen_columns[_]))
        # 2016
        for _ in range(len(sixteen_columns)):
            if sixteen_columns[_] not in eleven_columns:
                # update 2016 bag w/ index & instance
                _from_2016_non_coexist_2011.append((_,sixteen_columns[_]))

    # let us know if 2015 != 2016
    else:
        raise Exception('ERROR\nfifteen_columns != sixteen_columns\nERROR')


    '''2016 vs 2017 and reverse'''
    # store non coexist 2016 vs 2017
    _from_2016_non_coexist_2017 = []
    # store non coexist 2017 vs 2016
    _from_2017_non_coexist_2016 = []

    # from 2017 to 2016
    for _ in range(len(seventeen_columns)):
        if seventeen_columns[_] not in eleven_columns:
            # update 2017 bag w/ index & instance
            _from_2017_non_coexist_2016.append((_,seventeen_columns[_]))
    # from 2016 to 2017
    for _ in range(len(sixteen_columns)):
        if sixteen_columns[_] not in seventeen_columns:
            # update 2016 bag w/ index & instance
            _from_2016_non_coexist_2017.append((_,sixteen_columns[_]))

    '''2011 vs 2015/2016/2017'''
    # count non coexist 2011 vs 2015
    non_coexist_2015_count = len(non_coexist_2015)
    # count non coexist 2011 vs 2016
    non_coexist_2016_count = len(non_coexist_2016)
    # count non coexist 2011 vs 2017
    non_coexist_2017_count = len(non_coexist_2017)
    # count non coexist 2011 vs each in ALL
    non_coexist_count = non_coexist_2015_count + non_coexist_2016_count + non_coexist_2017_count
    # identify unique non coexistances
    u_non_coexist = set(non_coexist_2015 + non_coexist_2016 + non_coexist_2017)
    # count the number of unique non coexist 2011 vs each in All
    u_non_coexist_count = len(u_non_coexist)

    '''2016/2017 vs 2011'''
    # count non coexist 2016 vs 2011
    _from_2016_non_coexist_2011_count = len(_from_2016_non_coexist_2011)
    # count non coexist 2017 vs 2011
    _from_2017_non_coexist_2011_count = len(_from_2017_non_coexist_2011)
    # identify unique non coexistances (exist in 2016 or 2017 but not in 2011)
    r_u_non_coexist = set(_from_2016_non_coexist_2011 + _from_2017_non_coexist_2011)
    # count the number of unique non coexist 2011 vs each in All
    r_u_non_coexist_count = len(r_u_non_coexist)

    '''2016 & 2017'''
    # count non coexist 2016 vs 2017
    _from_2016_non_coexist_2017_count = len(_from_2016_non_coexist_2017)
    # count non coexist 2017 vs 2016
    _from_2017_non_coexist_2016_count = len(_from_2017_non_coexist_2016)

    # 2011 vs
    print(f'forward\n2011 vs 2015\nnon_coexist_2015_count = {non_coexist_2015_count} /{len(fifteen_columns)}\n'
            f'2011 vs 2016\nnon_coexist_2016_count = {non_coexist_2016_count} /{len(sixteen_columns)}\n'
            f'2011 vs 2017\nnon_coexist_2017_count = {non_coexist_2017_count} /{len(seventeen_columns)}\n'
            f'2011 vs all\ninstances not existing in all  = {u_non_coexist_count} / {len(eleven_columns)} possible\n')
    # vs 2011
    print(f'backward\n2016 vs 2011\n_from_2016_non_coexist_2011_count = {_from_2016_non_coexist_2011_count} /{len(sixteen_columns)}\n'
            f'2017 vs 2011\n_from_2017_non_coexist_2011_count = {_from_2017_non_coexist_2011_count} /{len(sixteen_columns)}\n'
            f'ALL vs 2011\nunique non coexist 2011 vs each in All = {r_u_non_coexist_count}\n')
    # 2016/2017 & 2017/2016
    print(f'2016 & 2016\n2016 vs 2017\n_from_2016_non_coexist_2017_count = {_from_2016_non_coexist_2017_count} /{len(sixteen_columns)}\n'
            f'2017 vs 2016\n_from_2017_non_coexist_2016_count = {_from_2017_non_coexist_2016_count} /{len(seventeen_columns)}\n')
# col_name_eda

- ***notes***:
    - 108 columns (distinctions) of 328
- ***possible***:
    - take out non-coexisting columns from all dataframes
        - then compare
            - would result in loss of ~1/3+ of columns
    - try to adjust
        - find what would be
            - seems possible for columns like the following
                - 10, 'Percent; SEX AND AGE - Male'
                - 96, 'Estimate; SEX AND AGE - Female'
                - 248, 'Estimate; RACE - Asian'
            - but more difficult for columns like
                - 11, 'Percent Margin of Error; SEX AND AGE - Male'
                - 97, 'Margin of Error; SEX AND AGE - Female'
                - 323, 'Percent Margin of Error; HISPANIC OR LATINO AND RACE - Not Hispanic or Latino - Two or more races - Two races excluding Some other race, and Three or more races')
                    - though ?could exclue this when calculating change?
        
- ***actions***:
    - time is becoming a more relevant an issue, MVP due today

In [ ]:
# make dataframe for each zip code, should contain row from each year
def frame_per_zip(dataframes):
    '''
    take dataframe for each year set is available
    find common zipcode for each year
    make dataframe for that zipcode
        containing each year's measurements
    note:
        dataframes must be same length
    '''
    # check length of dataframes is same
    for _ in range(len(dataframes)):
        # done to ensure each zip code has same represnetation 
        if len(dataframes[_-1]) != len(dataframes[_]):
            # stop if we have different lengths
            raise Exception(f'len(dataframe[{-_}]) != len(dataframe[{_}])')
        # also check that dataframes have same Id
        random_samples = [random.randint(1,int(len(dataframes[-_])/2)),
                          random.randint(1,int(len(dataframes[-_]))),
                          random.randint(int(len(dataframes[-_])/2),len(dataframes[-_]))]
        for sample in random_samples:
            # pull Id from df to compare
            if dataframes[_]['Id'][sample] != dataframes[_-2]['Id'][sample]:
                # stop if they don't match
                raise Exception(f"NON MATCHING Id\n{dataframes[_].Id[sample]}\nERROR\n{dataframes[_-2].Id[sample]}\n")
    
    # now we can get to work
    mini_dfs = []
    for i in range(len(dataframes[0])):
        mini_df = pd.DataFrame(index=['2011','2012','2013','2014','2015','2016','2017'], columns=dataframes[0].columns)
        for _ in range(len(dataframes)):
            q = dataframes[_]
            print(q)
            # mini_df[_] = q.index(i)
        mini_dfs.append(mini_df)
    
    print(len(mini_df))
        
    return mini_dfs
        
    

In [ ]:
# yrs = [sy2k11,sy2k12,sy2k13,sy2k14,sy2k15,sy2k16,sy2k17]
# z = frame_per_zip(yrs)

In [ ]:
for year in yrs:
    print (year.columns)

In [ ]:
# mini_df = pd.DataFrame(index=['2011','2012','2013','2014','2015','2016','2017'], columns=['a','b','c','d','e'])
# mini_df[1:2]

In [ ]:
# dataframes = [y2k11,y2k12,y2k13,y2k14,y2k15,y2k16,y2k17]
# mini_df = pd.DataFrame(index=['2011','2012','2013','2014','2015','2016','2017'], columns=dataframes[0].columns)
# mini_df.iloc[0] = 'lol'
# mini_df.iloc[0]
# mini_df

In [ ]:
'''# load 2000 data
y2k = pd.read_csv( a , low_memory=False )
# load 2010 data
y2k10 = pd.read_csv( b , low_memory=False )

# 2000 Census
b = y2k.copy()
# 2010 Census
o = y2k10.copy()

# reset 2000 columns to current 0th row values
b.columns = b.iloc[0]
# new 2000 dataframe without row where values are from
b = b[1:]
# reset index
b = b.reset_index()

# reset 2010 columns to current 0th row values
o.columns = o.iloc[0]
# new 2010 dataframe without row where values are from
o = o[1:]
# reset index
o = o.reset_index()

# identify zip codes from 2000 .Geography (last 5 chars of string)
zip_2000_codes = [q[-5:] for q in b.Geography]  # ValueError: invalid literal for int() with base 10: '006HH'
# identify zip codes from 2010 .Geography (last 5 chars of string)
zip_2010_codes = [q[-5:] for q in o.Geography]

# from 2000.Geography , instance is not seen in 2010.Geography  -- sample: zip_code = (2, 'c')
in_2000_but_not_2010_from_2000 = [zip_code for zip_code in enumerate(zip_2000_codes) if zip_code[1] not in zip_2010_codes]
# from 2010.Geography , instance is not seen in 2000.Geography  -- sample: zip_code[1] = 'c'
in_2010_but_not_2000_from_2010 = [zip_code for zip_code in enumerate(zip_2010_codes) if zip_code[1] not in zip_2000_codes]

# from 2000.Geography , instance is seen in 2010.Geography
in_2000_and_2010_from_2000 = [zip_code for zip_code in enumerate(zip_2000_codes) if zip_code[1] in zip_2010_codes]
# from 2010.Geography , instance is seen in 2000.Geography
in_2010_and_2000_from_2010 = [zip_code for zip_code in enumerate(zip_2010_codes) if zip_code[1] in zip_2000_codes]

# index of objects coexisting in 2000 and 2010
of_2000_indexes = [i for i,j in in_2000_and_2010_from_2000]
# index of objects coexisting in 2010 and 2000 
of_2010_indexes = [i for i,j in in_2010_and_2000_from_2010]
# ^note: these are different lists, if took j instead of i, then would be same list
if [j for i,j in in_2000_and_2010_from_2000] != [j for i,j in in_2010_and_2000_from_2010]:
    # like is seen here, j for j == True
    raise Exception(f'FLAWED ASSUMPTION , [j for i,j in 2000] != [j for i,j in 2010]\n'
                    f'len {len(in_2000_and_2010_from_2000)} {len(in_2010_and_2000_from_2010)}')
# however i for i == False
if of_2000_indexes == of_2010_indexes:
    # cheers
    raise Exception('FLAWED ASSUMPTION , of_2000_indexes != of_2010_indexes\n'
                    f'len y2k {len(of_2000_indexes)} 2k10 {len(of_2010_indexes)}')  

# thin 2000 to shared geo
b = b.iloc[of_2000_indexes]
# thin 2010 to shared geo
o = o.iloc[of_2010_indexes]'''
pass

In [ ]:
"""play df"""
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],'Age':[28,34,29,42], 'Weight':[128,134,129,142]}
df = pd.DataFrame(data)
df

In [ ]:
"""play list"""
a = [1,2,3,3,5,69,1,2,7,9,2]
b = [1,2,3]
c = [10,20,50,69]
d = set(a+b+c)
# d = set(d)
d

In [ ]:
from pandas.plotting import scatter_matrix
scatter_matrix(y2k11.set_index('Geography'),  figsize=(10, 10), diagonal='kde')

In [ ]:
q = [1,2,3,'4',5,6,7,2,6,1,9,1,4,1]
z = 0
while z ==0:
    for x in q:
        if isinstance(x,str):
            z+=1
        print(f'z={z}\nx={x}')

In [ ]:
old = pd.DataFrame({'A' : [4,5], 'B' : [10,20], 'C' : [100,50], 'D' : [-30,-50]})
old

In [ ]:
new = old[['A', 'C', 'D']].copy()
new

In [ ]:
'''
12-words
    combining census datasets to predict long term population trends (scalable)
yesterday: 
    mvp
today: 
    get a score
    enlarge dataset for long term trends
        e.g. historical migration trends vs census vs acs 5-yr /vs& acs 1-yr 
    explore other ml and sexy 
        if no time, set up for tomorrow
blocks:
    extracting data, usual
'''